### PR TITLE
[18NY] Implement privates

### PR DESCRIPTION
**Pull request by crericha**
### 18NY changes

- Implemented privates (except PA coal)
- Some operating round logic to test the privates

### Common code changes

- Refactored train_discount ability handling to support privates with multiple train_discount abilities
- Refactored tile_discount ability logging to use the spender name and not the private name when invoked from the special_tile step.
- Decoupled the special_token step's dependency on the special_track step.

### CODE REVIEW

**Comment by crericha** [comment]
Tested against all games in the 4/27 db and there were no failures.

File: lib/engine/ability/train_discount.rb [hunk 1]
```diff
@@ -7,10 +7,10 @@ module Ability
     class TrainDiscount < Base
       attr_reader :discount, :trains, :closed_when_used_up
 
-      def setup(discount:, trains:, closed_when_used_up: true)
+      def setup(discount:, trains:, closed_when_used_up: nil)
         @discount = discount
         @trains = trains
-        @closed_when_used_up = closed_when_used_up
+        @closed_when_used_up = closed_when_used_up || false
```
Comment (tobymao) [inline lib/engine/ability/train_discount.rb:13]: not necessary as nil is falsey

File: lib/engine/game/base.rb [hunk 2]
```diff
@@ -1219,7 +1219,7 @@ def graph_for_entity(_entity)
         @graph
       end
 
-      def upgrade_cost(tile, hex, entity)
+      def upgrade_cost(tile, hex, entity, spender)
```
Comment (tobymao) [inline lib/engine/game/base.rb:1222]: could be less change if you did 

entity, spender = nil

spender ||= entity

File: lib/engine/game/g_18_mex/step/track.rb [hunk 3]
```diff
@@ -37,7 +37,10 @@ def can_lay_tile?(entity)
             return super unless entity.minor?
 
             home_hex = @game.hex_by_id(entity.coordinates)
-            return @game.upgrade_cost(home_hex.tile, home_hex, entity) <= entity.cash if home_hex.tile.color == :white
+            if home_hex.tile.color == :white
+              return @game.upgrade_cost(home_hex.tile, home_hex, entity,
```
Comment (tobymao) [inline lib/engine/game/g_18_mex/step/track.rb:41]: one line

File: lib/engine/game/g_18_ny/game.rb [hunk 4]
```diff
@@ -166,6 +212,44 @@ def can_par?(corporation, _parrer)
         def can_hold_above_limit?(_entity)
           true
         end
+
+        # Operating round logic
+
+        def operating_order
+          minors, majors = @corporations.select(&:floated?).sort.partition { |c| c.type == :minor }
```
Comment (tobymao) [inline lib/engine/game/g_18_ny/game.rb:219]: why not just sort by [floated?, type.minor ? 0 : 1]

File: lib/engine/game/g_18_ny/game.rb [hunk 5]
```diff
@@ -166,6 +212,44 @@ def can_par?(corporation, _parrer)
...
+          minors + majors
+        end
+
+        def tile_lay(_hex, old_tile, _new_tile)
+          return unless old_tile.icons.any? { |icon| icon.name == ERIE_CANAL_ICON }
+
+          @log << "#{@erie_canal_private.name}'s revenue reduced from #{format_currency(@erie_canal_private.revenue)}" \
+                  " to #{format_currency(@erie_canal_private.revenue - 10)}"
+          @erie_canal_private.revenue -= 10
+          return unless @erie_canal_private.revenue.zero?
```
Comment (tobymao) [inline lib/engine/game/g_18_ny/game.rb:229]: return if revenue.positive

File: lib/engine/game/g_18_ny/step/stagecoach_exchange.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          end
+
+          def actions(_entity)
+            return %w[choose pass] if active?
+
+            []
+          end
+
+          def active?
+            return false if @game.stagecoach_token.nil?
```
Comment (tobymao) [inline lib/engine/game/g_18_ny/step/stagecoach_exchange.rb:21]: return false unless stagecoach_token

File: lib/engine/game/g_18_ny/step/stagecoach_exchange.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            return false if @passed && !@game.privates_closed
+
+            corporation = @game.stagecoach_token&.corporation
+            # Remove token if privates are closed and it can't be exchanged
+            remove_stagecoach_token if @game.privates_closed && (corporation.nil? || !corporation.next_token)
```
Comment (tobymao) [inline lib/engine/game/g_18_ny/step/stagecoach_exchange.rb:26]: !corporation || !corporation.next_token

!corporation&.next_token

File: lib/engine/step/special_buy_train.rb [hunk 8]
```diff
@@ -23,30 +23,32 @@ def blocks?
...
         pass! unless can_buy_train?(corporation)
       end
 
-      def ability(entity)
+      def ability(entity, train: nil)
         return unless entity&.company?
 
-        @game.abilities(entity, :train_discount, time: 'buying_train')
+        @game.abilities(entity, :train_discount, time: 'buying_train') do |ability|
+          return ability if train.nil? || ability.trains.include?(train.name)
```
Comment (tobymao) [inline lib/engine/step/special_buy_train.rb:48]: !train

File: lib/engine/step/tracker.rb [hunk 9]
```diff
@@ -243,25 +243,27 @@ def remove_border_calculate_cost!(tile, entity)
...
-            (!a.hexes || a.hexes.include?(hex.name))
-        end
-        discount = ability&.discount || 0
-
-        if discount.positive?
-          @log << "#{entity.name} receives a discount of #{@game.format_currency(discount)} from #{ability.owner.name}"
+      def border_cost_discount(entity, spender, border, cost, hex)
+        entity.all_abilities.each do |a|
+          next if (a.type != :tile_discount) ||
+            a.terrain.nil? ||
```
Comment (tobymao) [inline lib/engine/step/tracker.rb:254]: !a.terrain

File: lib/engine/game/g_18_ny/game.rb (hunk 4 above)
Comment (crericha) [reply lib/engine/game/g_18_ny/game.rb:219]: Not sorted on floated?. It is sorted on operating order, which is defined within corporation.

Comment (tobymao) [reply lib/engine/game/g_18_ny/game.rb:219]: ah i see nvm then

**Pull request merged**